add ieee802154 and entropy mcux driver and necessary changes

andrei-menzopol opened the pull request:
### Description 

drivers: entropy: add mcxw trng
drivers: ieee802154: add mcxw ieee802154 driver
dts: nxp: add ieee and trng bindings and nodes
soc: nxp: mcxw: update conditions to account for ieee driver
samples: echo_client echo_server: add frdm_mcxw7 board overlays
samples: echo_client echo_server: update tests
doc: frdm_mcxw71: update documentation
modules: hal_nxp: update commit to PR 491

### Testing

- Compiled `echo_server`/`echo_client` with: `west build -p always -b frdm_mcxw71 samples/net/sockets/echo_client -- -DEXTRA_CONF_FILE=overlay-ot.conf`
- Created a thread network with it and joined another thread device
- Checked that the thread communication is working (with a 15.4 thread wireshark capture)
- Was able to see the neighbors and ping

George-Stefan commented:
https://github.com/zephyrproject-rtos/zephyr/pull/82803

axelnxp commented on drivers/ieee802154/Kconfig.mcxw:
```diff
@@ -0,0 +1,15 @@
+
+menuconfig IEEE802154_MCXW
+	bool "NXP MCXW series IEEE 802.15.4 Driver"
+	default y
```
this will need a dependency to DTS, like:
`DT_HAS_NXP_MCXW_IEEE802154_ENABLED`

andrei-menzopol replied:
Thank you

axelnxp commented on drivers/ieee802154/Kconfig.mcxw:
```diff
@@ -0,0 +1,15 @@
+	bool "NXP MCXW series IEEE 802.15.4 Driver"
+	default y
+
+config CONFIG_IEEE802154_MCXW_INIT_PRIO
```
This should be under the `IEEE802154_MCXW` menuconfig, so
```
if IEEE802154_MCXW
config CONFIG_IEEE802154_MCXW_INIT_PRIO
...
endif
```

axelnxp commented on west.yml:
```diff
@@ -203,7 +203,7 @@ manifest:
         - hal
     - name: hal_nxp
-      revision: 0ac830233092247c26f5dd01a07b0a484532ea4c
+      revision: pull/491
```
thought the syntax was:
```
pull/<id>/head
```

axelnxp reviewed (changes requested):
I know it's still a draft, but suggesting some changes for the final version

axelnxp commented on soc/nxp/mcx/mcxw/CMakeLists.txt:
```diff
@@ -15,3 +15,5 @@ zephyr_include_directories(.)
 
 zephyr_linker_sources_ifdef(CONFIG_BT RAM_SECTIONS sections.ld)
+
+zephyr_linker_sources_ifdef(CONFIG_IEEE802154 RAM_SECTIONS sections.ld)
```
we might just regroup this under a single if
```
if(CONFIG_BT OR CONFIG_IEEE802154)
...
endif()
```

andrei-menzopol replied:
Done, thanks

andrei-menzopol commented:
I would like to merge these PRs for 4.1 RC1 if possible

rlubos commented on samples/net/sockets/echo_client/sample.yaml:
```diff
@@ -70,6 +70,14 @@ tests:
   sample.net.sockets.echo_client.nrf_802154:
     extra_args: EXTRA_CONF_FILE="overlay-802154.conf"
     platform_allow: nrf52840dk/nrf52840
+  sample.net.sockets.echo_client.mcxw71_openthread:
```
Can we have a single `sample.net.sockets.echo_client.openthread` entry instead and just list the supported platforms in the `platform_alow:` section? It doesn't really make sense to multiply those build variants given the only difference is the platform.

andrei-menzopol replied:
Done, please check.
Thanks

rlubos replied:
Thanks

axelnxp commented on dts/arm/nxp/nxp_mcxw71.dtsi:
```diff
@@ -27,14 +27,14 @@
 	stcm1: system_memory@1a000 {
 		compatible = "zephyr,memory-region","mmio-sram";
-		reg = <0x1a000 DT_SIZE_K(8)>;
+		reg = <0x1a000 DT_SIZE_K(4)>;
```
this was originally describind the STCM5 which is powered by VBAT in the always domain, and it's a 8K ram bank.
I'm not sure this change is valid, can you explain ?

andrei-menzopol replied:
I had to do this change because `samples/net/sockets/echo_server`'s SRAM wouldn't fit into the 104KB stcm0. We should find another solution.

George-Stefan replied:
decrease some of OT stack configs

decsny replied:
@andrei-menzopol you can make a DT overlay for MCXW71 on this sample if needed, I agree with @axelnxp not to change it here

andrei-menzopol replied:
Removed this, addre board overlays

axelnxp commented on drivers/entropy/entropy_mcxw.c:
```diff
@@ -0,0 +1,71 @@
+
+	k_sem_take(&entropy_mcxw_data.sem_lock, K_FOREVER);
+
+    if (CRYPTO_InitHardware() != kStatus_Success) {
```
there's a weird indentation going on, make sure to align everything correctly using tabs and not spaces

andrei-menzopol replied:
Done, ran clang-format

axelnxp commented on drivers/ieee802154/CMakeLists.txt:
```diff
@@ -15,6 +15,8 @@ zephyr_library_sources_ifdef(CONFIG_IEEE802154_CC2520		ieee802154_cc2520.c)
 zephyr_library_sources_ifdef(CONFIG_IEEE802154_DW1000		ieee802154_dw1000.c)
 zephyr_library_sources_ifdef(CONFIG_IEEE802154_KW41Z		ieee802154_kw41z.c)
 zephyr_library_sources_ifdef(CONFIG_IEEE802154_MCR20A		ieee802154_mcr20a.c)
+zephyr_library_sources_ifdef(CONFIG_IEEE802154_MCXW			ieee802154_mcxw.c)
```
you can list all sources into a single statement, but this might be against zephyr convention, @decsny what do you think ?

dleach02 replied:
do you mean all _MCXW sources? Searching Zephyr codebase I don't see any examples of having multiple sources in a single statement.

decsny replied:
There is no problem with doing that nor any convention. And there are examples. It is supported. I agree with @axelnxp to do that but this is a nit obviously

andrei-menzopol replied:
Done

axelnxp commented on drivers/ieee802154/ieee802154_mcxw.c:
Please make sure to only use tabs in your sources and not mix tabs and spaces, the indentation looks weird.

andrei-menzopol replied:
Done, ran clang-format

ankuns reviewed (dismissed):
No caveats from my point of view, however please think about my approval rather as "it does not break what has already been there". When it comes to the content, I'm leaving the review for those interested in chips from NXP.

axelnxp commented on drivers/ieee802154/ieee802154_mcxw.c:
```diff
@@ -0,0 +1,1334 @@
+	OSA_InterruptDisable();
+
+	/* Get new 32bit HW timestamp */
+	uint64_t u64HwTimestampUs_new = (uint64_t)TM_GetTimestamp();
```
@decsny is there a zephyr api that could be used directly to read timestamps ? and redirect this to LPTMR ? I don't know if that's supported

andrei-menzopol replied:
used lptmr counter instead

George-Stefan commented on drivers/ieee802154/ieee802154_mcxw.c:
```diff
@@ -0,0 +1,1346 @@
+
+	u64HwTimestampUs_new = counter_ticks_to_us(mcxw_ctx.counter, u32Ticks);
+
+	OSA_InterruptDisable();
```
Move it before counter_get_value() or remove it

andrei-menzopol replied:
removed

Commit:
drivers: entropy: add nxp ele trng

Add NXP ELE TRNG driver

Signed-off-by: Andrei Menzopol <andrei.menzopol@nxp.com>

Commit:
drivers: ieee802154: add mcxw ieee802154 driver

Add mcxw ieee802154 driver
Fix compliance errors for added files.
Update function names to snake_case style.

Signed-off-by: Andrei Menzopol <andrei.menzopol@nxp.com>

Commit:
dts: nxp: add ieee and trng bindings and nodes

Add ieee802154 and trng bindings and nodes

Signed-off-by: Andrei Menzopol <andrei.menzopol@nxp.com>

Commit:
soc: nxp: mcxw: update conditions to account for ieee driver

Add ram section for ieee driver.
Use config for ieee driver.

Signed-off-by: Andrei Menzopol <andrei.menzopol@nxp.com>

Commit:
samples: echo_client echo_server: add frdm_mcxw7 board overlays

Add frdm_mcxw7 board conf and overlay files for echo_client and
echo_server sample applications.

Signed-off-by: Andrei Menzopol <andrei.menzopol@nxp.com>

Commit:
samples: echo_client echo_server: update tests

Group openthread tests under the same case and add frdm_mcxw71.

Signed-off-by: Andrei Menzopol <andrei.menzopol@nxp.com>

Commit:
doc: frdm_mcxw71: update documentation

Update documentation for dynamic NBU.
Add openthread application building examples.

Signed-off-by: Andrei Menzopol <andrei.menzopol@nxp.com>

rlubos commented on samples/net/sockets/echo_client/boards/frdm_mcxw71.conf:
```diff
@@ -0,0 +1,42 @@
+CONFIG_COUNTER=y
+
+# Disable TCP and IPv4 (TCP disabled to avoid heavy traffic)
+CONFIG_NET_TCP=n
```
Board overlays shouldn't copy generic configuration, just provide board-specific tweaks. IMHO the build should still rely on `overlay-ot.conf` in general, just provide config adjustments here.

andrei-menzopol replied:
If I compile with `-DCONF_FILE="prj.conf overlay-ot.conf"` it doesn't take `boards/frdm_mcxw71.conf` automatically anymore.

andrei-menzopol replied:
I believe I have to pass the `boards/frdm_mcxw71.conf` too in `DCONF_FILE`
https://docs.zephyrproject.org/latest/build/kconfig/setting.html#the-initial-configuration

andrei-menzopol replied:
I updated the conf file

rlubos replied:
> If I compile with -DCONF_FILE="prj.conf overlay-ot.conf" it doesn't take boards/frdm_mcxw71.conf automatically anymore.

That's because `DCONF_FILE` overrides the entire configuration, to apply overlays you should use `EXTRA_CONF_FILE` - that way Zephyr's build system applies configurations from all sources.
So building like `west build -b frdm_mcxw71 -- -DEXTRA_CONF_FILE=overlay-ot.conf` will merge all needed configs (prj.conf, overlay and board-specific overlay):
![image](https://github.com/user-attachments/assets/7778d2f9-2336-4137-9a3a-1fc5f23ecd56)

rlubos commented on samples/net/sockets/echo_client/sample.yaml:
```diff
@@ -78,10 +78,20 @@ tests:
     filter: CONFIG_FULL_LIBC_SUPPORTED and not CONFIG_NATIVE_LIBC
+  sample.net.sockets.echo_client.mcxw_openthread:
+    extra_args:
+      - EXTRA_CONF_FILE=./boards/frdm_mcxw71.conf
+      - DTC_OVERLAY_FILE=./boards/frdm_mcxw71.overlay
```
That shouldn't be really needed, board overlays should be applied automatically by the build system.

andrei-menzopol replied:
Removed the `DTC_OVERLAY_FILE`. Thank you

rlubos commented on drivers/ieee802154/ieee802154_mcxw_utils.c:
```diff
@@ -0,0 +1,350 @@
+	return NULL;
+}
+
+void SetCslIe(uint8_t *pdu, uint16_t lenght, uint16_t period, uint16_t phase)
```
General note - Zephyr code should be written in snake_case: https://docs.zephyrproject.org/latest/contribute/style/code.html

andrei-menzopol replied:
Done. Thanks

rlubos replied:
Is it? I see all the functions still use PascalCase.

andrei-menzopol replied:
Yes, sorry. Now it's done

rlubos reviewed (commented):
Commits with compliance/style fixes should be squashed.
+1 for sample and 15.4 changes, didn't review entropy driver.

cfriedt commented on boards/nxp/frdm_mcxw71/doc/index.rst:
```diff
@@ -194,6 +213,19 @@ Please consider changing ``COMxx`` on Windows or ``ttyxx`` on Linux to the seria
 
 The NBU file can be found in : ``<zephyr workspace>/modules/hal/nxp/zephyr/blobs/mcxw71/mcxw71_nbu_ble.sb3``
 
+For dynamic (bluetooth + ieee802154) NBU:
+.. code-block:: console
+
+   # On Windows
+   blhost.exe -p COMxx -- receive-sb-file mcxw71_nbu_ble_15_4_dyn.sb3
+
+   # On Linux
+   ./blhost -p /dev/ttyxx -- receive-sb-file mcxw71_nbu_ble_15_4_dyn.sb3
+
+Please consider changing ``COMxx`` on Windows or ``ttyxx`` on Linux to the serial port used by your board.
+
+The NBU file can be found in : ``<zephyr workspace>/modules/hal/nxp/zephyr/blobs/mcxw71/mcxw71_nbu_ble_15_4_dyn.sb3``
```
There is a helper for doing this using tabs, which renders to a slightly cleaner interface.

https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/modules/thrift/hello/README.rst?plain=1#L65

andrei-menzopol replied:
Done, thank you

cfriedt reviewed (commented):
Just one suggestion to improve sample docs.

Otherwise, looks good to me 👍

andrei-menzopol commented:
I think the PR is ready. Can I get some reviews please? Thank you

jukkar commented on drivers/entropy/entropy_mcxw.c:
```diff
@@ -0,0 +1,63 @@
+	if (CRYPTO_InitHardware() != kStatus_Success) {
+		goto exit;
+	}
+	if (sss_sscp_rng_context_init(&g_sssSession, &ctx, 0u) != kStatus_SSS_Success) {
+	} else if (sss_sscp_rng_get_random(&ctx, buf, len) != kStatus_SSS_Success) {
+	} else if (sss_sscp_rng_free(&ctx) != kStatus_SSS_Success) {
```
What is the point of these empty checks, you should be able to create a proper `if` check here?

decsny commented on dts/bindings/ieee802154/nxp,mcxw-ieee802154.yaml:
```diff
@@ -0,0 +1,8 @@
+
+description: NXP MCXW71 IEEE 802.15.4 node
+
+compatible: "nxp,mcxw-ieee802154"
```
similar comment, do not put mcxw in compatible string, in case this radio IP is also on another part, is there a name of the IP?

andrei-menzopol replied:
Currently the SDK Phy API that is used by this driver only supports two boards MCXW71 and MCXW72 so I believe that this name is ok in this case

decsny replied:
we really have a hard rule not to put soc names in the IP compatible strings. for now i will not block this PR merging over it, since it's not a node a board would define, but really needs to be updated later

George-Stefan replied:
what do you propose?
We can update it when introducing the mcxw72 ieee 802.15.4 driver?

George-Stefan replied:
The fw runs on the NBU core. You can't access directly the radio hw

decsny commented on drivers/ieee802154/ieee802154_mcxw.h:
```diff
@@ -0,0 +1,148 @@
+
+	/* CSL sample time in microseconds */
+	uint32_t csl_sample_time;
+
```
nit remove all these newlines, its not consistent with zephyr codebase

andrei-menzopol replied:
Done. Thanks

decsny commented on boards/nxp/frdm_mcxw71/doc/index.rst:
need to update the supported feature table?

andrei-menzopol replied:
Done. Thanks

andrei-menzopol commented on dts/bindings/rng/nxp,mcxw-trng.yaml:
```diff
@@ -0,0 +1,8 @@
+
+description: NXP MCXW TRNG (True Random Number Generator)
+
+compatible: "nxp,mcxw-trng"
```
I agree, ELE (EdgeLock secure enclave) is more generic. I will make the updates

andrei-menzopol replied:
Done

George-Stefan commented:
> Just one suggestion to improve sample docs.
> 
> Otherwise, looks good to me 👍

@cfriedt could you please review again the PR.
All comments were addressed

George-Stefan commented:
> Commits with compliance/style fixes should be squashed. +1 for sample and 15.4 changes, didn't review entropy driver.

@rlubos  could you please review again the PR.
All comments were addressed

ceolin reviewed (dismissed):
entropy side looks good to me.

George-Stefan commented on west.yml:
```diff
@@ -203,7 +203,7 @@ manifest:
       groups:
         - hal
     - name: hal_nxp
-      revision: 07a87b69098d8e0c9b93355bde411e72427867ba
```
ok

George-Stefan committed:
west.yml: update hal_nxp to support new features on MCXW71

- add MCXW71 NBU combo BLE LL and IEEE 802.15.4 PHY firmware blob
- fix elemu driver: undef BIT if already defined
- add mcux-secure-subsystem middleware
- add ieee_802.15.4 middleware

Signed-off-by: George Stefan <george.stefan@nxp.com>

George-Stefan commented:
> entropy side looks good to me.

@ceolin, @rlubos , @cfriedt, @mmahadevan108 
NXP HAL PR was merged so this is the final update of this PR

dleach02 commented:
@ceolin can we get your review?

dleach02 commented:
TSC approved the blob RFC
